Update outdated testing documentation (use pytest)

SubaruArai opened the pull request:
The CONTRIBUTING.md's "How to submit pull requests" specified nosetests, which is no longer used. (see: commit id f27dfbf, github pr #36201)
This commit updates the documentation to use pytest instead, and also encourages the use of a virtual environment.

Config files are also updated to ignore the virtual environment.

Commit:
Update outdated testing documentation (use pytest)

The CONTRIBUTING.md's "How to submit pull requests" specified nosetests,
which is no longer used. (see: commit id f27dfbf, github pr #36201)
This commit updates the documentation to use pytest instead, and also
encourages the use of a virtual environment.

Config files are also updated to ignore the virtual environment.

clalancette reviewed (changes requested):
I'll note that this will conflict with https://github.com/ros/rosdistro/pull/38125 .  There is more information in here than in that PR, but I think we should probably get that one in first and then rebase this on top.

Commit:
Merge branch 'master' into docs/use_venv

SubaruArai commented:
@clalancette merged, please check!

Commit:
Swap order in CONTRIBUTING.md for readability

Swapped the order of information presented in CONTRIBUTING.md -
"Testing/Test Dependencies".
New users likely care more about how to run tests than what it depends
on.

Commit:
Clarify how to run pytest in CONTRIBUTING.md

Added `pytest` command right after the dependency installation script

SubaruArai commented:
@clalancette ping

clalancette commented on CONTRIBUTING.md:
```diff
@@ -318,10 +337,17 @@ These tests require several dependencies that can be installed either from the R
 | unidiff        | python3-unidiff (Zesty and higher) | unidiff        |
 | yamllint       | yamllint                          | yamllint       |
```
This whole thing seems a bit redundant.  We tell the user 3 times to run `pytest` locally.

What I'm going to suggest instead is that we rewrite this to be simpler.  In particular, let's just add in one section called "Unit testing" or something like that.  In there, we say something like:

```
All of the dependencies for testing are listed in tests/requirements.txt.  It is recommended to use a virtual environment and pip to install the dependencies.
```

And then go on to give those instructions on how to install and run.  We can then remove the rest of the duplicated information here, so this is a lot shorter and more to the point.

SubaruArai replied:
lgtm

clalancette commented on CONTRIBUTING.md:
````diff
@@ -318,10 +337,17 @@ These tests require several dependencies that can be installed either from the R
+#### Quick testing
+
 There is a tool [scripts/check_rosdep](./scripts/check_rosdep.py) which will check most formatting errors such as alphabetization and correct formatting.
-It is recommended to run it before submitting your contribution.
+It will not check for all errors, but it is a quick way to check for common errors.
+
+Here is a one liner to check all files in the [rosdep](./rosdep) directory:
+```bash
+for f in ./rosdep/*.yaml; do echo "$f"; ./scripts/check_rosdep.py "$f"; done
+```
 
-For example, to check a change to `rosdep/base.yaml`:
+Or simply use `pytest` (although the output is quite verbose):
 ```bash
-python3 scripts/check_rosdep.py rosdep/base.yaml
+pytest test/rosdep_formatting_test.py
 ```
````
Honestly, I'm not sure we need this bit at all.  This is already covered by the general testing above, so maybe we can just drop this stanza completely?

SubaruArai replied:
Yeah, this was basically a leftover that I didn't know if I should've deleted.
I'm fine for dropping.

Commit:
update CONTRIBUTING.md: simplify testing section

SubaruArai commented:
@clalancette updated, please check!

Commit:
update CONTRIBUTING: clarify some wording

SubaruArai commented:
@clalancette ping

SubaruArai commented:
@clalancette ping

clalancette reviewed (approved):
Sorry for the delay here.  This looks good to me now.
